File: david/03/solution_part2.java
/*
 * Corrupted mul part 2
 *  do() .... dont() .... do() ...
 * 
 * match for pattern / dont() / do()

 */

import java.io.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
public class solution_part2 {
    private static String regex = "mul\\(([1-9][0-9]*),([1-9][0-9]*)\\)";
    private static String doString = "do()";
    private static String dontString = "don't()";
    private static String doRegex = "do\\(\\)";
    private static String dontRegex = "don't\\(\\)";
    private static String combinedRegex = regex + "|" + dontRegex + "|" + doRegex;

    public static void main(String[] args) throws FileNotFoundException {
        String line = new Scanner(new File(".//input.txt")).useDelimiter("\\Z").next();
        Pattern combinedPattern = Pattern.compile(combinedRegex);
        int result = 0;
        boolean enabled = true; // assuming enablement doesn't change between input lines
        Matcher combinedMatcher = combinedPattern.matcher(line);

        while (combinedMatcher.find()) {
            System.out.println(combinedMatcher.group());
            if (combinedMatcher.group().equals(doString)) {
                enabled = true;
            } else if (combinedMatcher.group().equals(dontString)) {
                enabled = false;
            } else if (enabled) {
                int left = Integer.parseInt(combinedMatcher.group(1));
                int right = Integer.parseInt(combinedMatcher.group(2));
                result += left * right;
            }
        }
        System.out.println("final result: " + result);
    }
}

File: koguchi/03/solution_part2.py
import re


with open('input.txt', 'r') as f:
    input_string = f.read()


pattern = r'mul\(\d+,\d+\)|do\(\)|don\'t\(\)'
matches = re.findall(pattern, input_string)
pattern_extract = r"mul\((\d+),(\d+)\)"

total = 0
do = True
for match in matches:
    print(match)
    if "don't" in match:
        do = False
    elif "do" in match:
        do = True
    else:
        m = re.match(pattern_extract, match)
        if m:
            a = int(m.group(1))
            b = int(m.group(2))
            if do:
                product = a * b
                total += product

print(total) # 82857512
